herd/X86_64Sem.ml

```diff
@@ -255,10 +262,35 @@ module
         M.mk_singleton_es
           (Act.Arch (X86_64.ArchAction.ClFlush (opt,a))) ii
 
-      let build_semantics _ ii =
+      let v2tgt =
+        let open Constant in
+        function
+        | M.A.V.Val(Label (_, lbl)) -> Some (B.Lbl lbl)
+        | M.A.V.Val (Concrete i) -> Some (B.Addr (M.A.V.Cst.Scalar.to_int i))
+        | _ -> None
+
+      let do_indirect_jump test bds i v =
+        match  v2tgt v with
+        | Some tgt -> M.unitT (B.Jump (tgt,bds))
+        | None ->
+           match v with
+           | M.A.V.Var(_) as v ->
+              let lbls = get_exported_labels test in
+              if Label.Full.Set.is_empty lbls then
+                M.unitT () >>! B.Exit
+              else
+                B.indirectBranchT v lbls bds
+        | _ -> Warn.fatal
+            "illegal argument for the indirect branch instruction %s \
+            (must be a label)" (X86_64.dump_instruction i)
+
+      let build_semantics test ii =
         let rec build_semantics_inner locked ii =
           match ii.A.inst with
           | X86_64.I_NOP -> B.nextT
+          | X86_64.I_RET as i ->
+            read_reg true X86_64.RIP ii
+            >>= do_indirect_jump test [] i
```

Comment:
I am puzzled, I thought that the `RET` instruction popped the return address from the stack.

Reply:
Apologies yes it appears to do that - so I need to add support for PUSH/POP for this to make sense in herd, and presumably work in litmus also?